drivers/counter/counter_nrfx_timer.c

```diff
@@ -55,112 +65,173 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+static u32_t get_top_value(struct device *dev)
+{
+	return nrf_timer_cc_read(get_nrfx_config(dev)->timer, TOP_CH);
+}
+
+static u32_t get_max_relative_alarm(struct device *dev)
 {
-	return nrfx_timer_capture_get(&get_nrfx_config(dev)->timer, TOP_CH);
+	return get_top_value(dev);
 }
 
-static u32_t counter_nrfx_get_max_relative_alarm(struct device *dev)
+static u32_t read(struct device *dev)
 {
-	return nrfx_timer_capture_get(&get_nrfx_config(dev)->timer, TOP_CH);
+	NRF_TIMER_Type *timer = get_nrfx_config(dev)->timer;
+
+	nrf_timer_task_trigger(timer,
+			       nrf_timer_capture_task_get(COUNTER_READ_CC));
+
+	return nrf_timer_cc_read(timer, COUNTER_READ_CC);
 }
 
-static u32_t counter_nrfx_read(struct device *dev)
+/* Return true if value equals 2^n - 1 */
+static inline bool is_power_of_2(u32_t val)
 {
-	return nrfx_timer_capture(&get_nrfx_config(dev)->timer,
-				  COUNTER_READ_CC);
+	return !(val & (val + 1));
 }
 
-/** @brief Calculate compare value.
- *
- * If ticks are relative then compare value must take into consideration
- * counter wrapping.
- *
- * @return Compare value to be used in TIMER channel.
- */
-static inline u32_t counter_nrfx_get_cc_value(struct device *dev,
-				const struct counter_alarm_cfg *alarm_cfg)
+static u32_t ticks_add(u32_t val1, u32_t val2, u32_t top)
 {
-	u32_t remainder;
-	u32_t cc_val;
-	u32_t ticks = alarm_cfg->ticks;
+	u32_t sum = val1 + val2;
 
-	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) != 0) {
-		return ticks;
+	if (likely(is_power_of_2(top))) {
+		sum = sum & top;
+	} else {
+		sum = sum > top ? sum - top : sum;
 	}
 
-	cc_val = counter_nrfx_read(dev);
-	remainder = counter_nrfx_get_top_value(dev) - cc_val;
+	return sum;
+}
+
+static u32_t ticks_sub(u32_t val, u32_t old, u32_t top)
+{
+	if (likely(is_power_of_2(top))) {
+		return (val - old) & top;
+	}
+
+	/* if top is not 2^n-1 */
+	return (val > old) ? (val - old) : val + top + 1 - old;
+
+}
+
+static void set_cc_int_pending(struct device *dev, u8_t chan)
+{
+	get_dev_data(dev)->cc_int_pending |= BIT(chan);
+	NRFX_IRQ_PENDING_SET(NRFX_IRQ_NUMBER_GET(get_nrfx_config(dev)->timer));
+}
+
+static int set_cc(struct device *dev, u8_t chan, u32_t val, u32_t flags)
+{
+	bool absolute = flags & COUNTER_ALARM_CFG_ABSOLUTE;
+	bool trigger_late = flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+	NRF_TIMER_Type  *reg = get_nrfx_config(dev)->timer;
+	nrf_timer_event_t evt = CH2EVT(chan);
+	u32_t top = counter_get_top_value(dev);
+	int err = 0;
+	u32_t prev_val;
+	u32_t now;
+	u32_t diff;
+	u32_t max_abs_val;
+
+	max_abs_val = top - get_dev_data(dev)->guard_period;
+	/* First take care of a risk of an event coming from CC being set to
+	 * next tick. Reconfigure CC to future (now tick is the furtherest
+	 * future).
+	 */
+	now = read(dev);
+	prev_val = nrf_timer_cc_read(reg, chan);
+	nrf_timer_cc_write(reg, chan, now);
+	nrf_timer_event_clear(reg, evt);
+
+	if (!absolute) {
+		trigger_late = val < (top / 2);
```

Comment:
Could you provide some comment explaining the above? Why are the relative ticks compared against half of the top value? And why is this `bool` value used in the shifting right below?

Reply:
added comment